File: app/src/main/java/org/mifos/selfserviceapp/utils/MFErrorParser.java
package org.mifos.selfserviceapp.utils;

import com.google.gson.Gson;

import org.mifos.selfserviceapp.models.mifoserror.MifosError;

import retrofit2.HttpException;
import rx.plugins.RxJavaPlugins;

public class MFErrorParser {

    public static final String LOG_TAG = "MFErrorParser";

    private static Gson gson = new Gson();

    public static MifosError parseError(String serverResponse) {
        return gson.fromJson(serverResponse, MifosError.class);
    }

    public static String errorMessage(Throwable throwableError) {
        String errorMessage = "";
        try {
            if (throwableError instanceof HttpException) {
                errorMessage = ((HttpException) throwableError).response().errorBody().string();
                errorMessage =  MFErrorParser.parseError(errorMessage).getErrors()
                        .get(0).getDefaultUserMessage();
            }
        } catch (Throwable throwable) {
            RxJavaPlugins.getInstance().getErrorHandler().handleError(throwable);
        }
        return errorMessage;
    }
}

File: androidApp/src/main/java/org/mifos/mobile/utils/MFErrorParser.kt
package org.mifos.mobile.utils

import com.google.gson.Gson
import io.reactivex.plugins.RxJavaPlugins
import org.mifos.mobile.core.model.entity.mifoserror.MifosError
import retrofit2.HttpException

object MFErrorParser {
    const val LOG_TAG = "MFErrorParser"
    private val gson = Gson()
    fun parseError(serverResponse: String?): MifosError {
        return gson.fromJson(serverResponse, MifosError::class.java)
    }

    @JvmStatic
    fun errorMessage(throwableError: Throwable?): String? {
        var errorMessage: String? = ""
        try {
            if (throwableError is HttpException) {
                errorMessage = throwableError.response()?.errorBody()?.string()
                errorMessage = parseError(errorMessage).errors[0].defaultUserMessage
            }
        } catch (throwable: Throwable) {
            RxJavaPlugins.getErrorHandler()
        }
        return errorMessage
    }
}
